### PR TITLE
Repository component preview

### PR DESCRIPTION
**Pull request by rkusan00**
- [x] Added frontend route for container preview 
- [x] Added backend route for container preview 
- [x] Connect routes to preview received JSON data

### CODE REVIEW

File: client/student/components/program/Activity/index.vue [hunk 1]
```diff
@@ -46,6 +45,11 @@ export default {
         : find(courseware, { subActivities: [{ id }] }).subActivities;
     }
   },
+  methods: {
+    async getContainer() {
```
Comment (vladimyr) [inline client/student/components/program/Activity/index.vue]: Why is this marked as `async`, you don't do `await` inside...

File: client/student/components/program/Activity/PreviewContainer.vue [hunk 2]
```diff
@@ -0,0 +1,25 @@
+<template>
```
Comment (vladimyr) [inline client/student/components/program/Activity/PreviewContainer.vue]: What is the point of having another component that is used exclusively as a wrapper for `<content-container>`? Why not using it directly?

File: client/student/router.js [hunk 3]
```diff
@@ -66,7 +67,14 @@ const router = new Router({
         props: parseParams
       }]
     }]
-  }, fallbackRoute]
+  },
+  {
+    path: '/previewComponent/:containerId',
+    name: 'preview',
+    component: PreviewComponent,
```
Comment (vladimyr) [inline client/student/router.js]: Just use `<content-container>` with different props? 🤔

File: server/preview/index.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
+'use strict';
+
+const router = require('express').Router();
+const fs = require('fs');
+var path = require('path');
```
Comment (vladimyr) [inline server/preview/index.js]: Hmm :bowtie: No `var`s allowed...

File: server/preview/index.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+const writeFile = promisify(fs.writeFile);
+const readFile = promisify(fs.readFile);
+const deleteFile = promisify(fs.unlink);
+
+router
+  .post('/', createPreview)
+  .get('/:containerId', fetchPreview);
+
+async function createPreview(req, res) {
+  const filePath = path.join(__dirname, req.body.id + '.json');
```
Comment (vladimyr) [inline server/preview/index.js]: Use template string for filename part creation:
```js
  const filePath = path.join(__dirname, `${req.body.id}.json`);
```

File: server/preview/index.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  await writeFile(filePath, JSON.stringify(req.body));
+  return res.redirect(`/#/previewComponent/${req.body.id}`);
+}
+
+async function fetchPreview(req, res) {
+  const { containerId } = req.params;
+  const filePath = path.join(__dirname, containerId + '.json');
```
Comment (vladimyr) [inline server/preview/index.js]: Same as above ⬆️

File: server/preview/index.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
+'use strict';
+
+const router = require('express').Router();
+const fs = require('fs');
+var path = require('path');
+const { promisify } = require('util');
+
+const writeFile = promisify(fs.writeFile);
+const readFile = promisify(fs.readFile);
```
Comment (vladimyr) [inline server/preview/index.js]: How about implementing `readJsonFile` & `writeJsonFile` instead? 😉

**Review by vladimyr — CHANGES_REQUESTED** [review]
Sorry for waiting, I'll be quicker next time 🏃 
Anyway, looks good, left few comments for you to address 👍

**Comment by zcuric** [comment]
What needs to be finished here? Any roadmap?

**Comment by zcuric** [comment]
@underscope

File: client/student/router.js [hunk 8]
```diff
@@ -66,14 +72,21 @@ const router = new Router({
         props: parseParams
       }]
     }]
-  }, fallbackRoute]
+  },
+  {
+    path: '/previewComponent/:containerId',
+    name: 'preview',
+    component: ContentContainer,
+    props: getContainer
```
Comment (kronicker) [inline client/student/router.js]:
```js
const getPreviewContainer = () => api.getPreview(params.containerId)

// ...

props: { getContainer: getPreviewContainer }
```

File: client/student/router.js [hunk 9]
```diff
@@ -66,14 +72,21 @@ const router = new Router({
         props: parseParams
       }]
     }]
-  }, fallbackRoute]
+  },
+  {
+    path: '/previewComponent/:containerId',
```
Comment (kronicker) [inline client/student/router.js]: `preview/:containerId`

**Review by kronicker — CHANGES_REQUESTED** [review]

File: client/student/api/content.js [hunk 10]
```diff
@@ -2,6 +2,9 @@ import { extractData } from '@/common/api/helpers';
 import request from '@/common/api/request';
 
 const url = {
+  preview: id => {
+    return `preview/${id}`;
+  },
```
Comment (kronicker) [inline client/student/api/content.js]: Inline

File: client/student/components/preview/Activity.vue [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+    </template>
+  </div>
+</template>
+
+<script>
+import ContentContainer from './common/ContentContainer';
+import get from 'lodash/get';
+import orderBy from 'lodash/orderBy';
+
+export default {
```
Comment (kronicker) [inline client/student/components/preview/Activity.vue:14]: Missing `name` option

File: client/student/components/preview/Activity.vue [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+  props: {
+    activity: { type: Object, required: true }
+  },
+  computed: {
+    contentContainers() {
+      const contentContainers = get(this.activity, 'containers', []);
+      const rules = [['type', 'position'], ['desc', 'asc']];
```
Comment (kronicker) [inline client/student/components/preview/Activity.vue]: Don't order by type

File: client/student/components/preview/index.vue [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+  </div>
+</template>
+
+<script>
+import Activity from './Activity';
+import api from '@/student/api/content';
+import CircularProgress from '@/student/components/common/CircularProgress';
+import Navbar from '@/student/components/common/Navbar';
+
+export default {
```
Comment (kronicker) [inline client/student/components/preview/index.vue:18]: Missing `name`

File: client/student/components/preview/index.vue [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+      activity: null,
+      isLoading: true
+    };
+  },
+  created() {
+    api.getPreview(this.contentPreviewId)
+      .then(activity => {
+        this.activity = activity;
+        this.isLoading = false;
+      });
```
Comment (kronicker) [inline client/student/components/preview/index.vue:36]: The loader will hang on error

File: client/student/components/preview/index.vue [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+    CircularProgress,
+    Navbar
+  }
+};
+</script>
+
+<style lang="scss" scoped>
+.container {
+  padding: 50px 0;
+}
```
Comment (kronicker) [inline client/student/components/preview/index.vue:49]: Why this padding?

File: client/student/components/preview/index.vue [hunk 16]
```diff
@@ -0,0 +1,53 @@
+<template>
+  <div>
+    <navbar />
+    <div class="container">
+      <div v-if="isLoading" class="loader-container">
+        <circular-progress :height="50" :width="50" />
+      </div>
```
Comment (kronicker) [inline client/student/components/preview/index.vue]: We have this in multiple places along with the styling. Extract it to a common component like `progress-bar` and reuse throughout.

File: client/student/components/preview/common/ContentContainer.vue [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+.content-container {
+  display: flex;
+  justify-content: center;
+  margin: 35px 0;
+
+  .columns {
+    flex: 1;
+  }
+}
+</style>
```
Comment (kronicker) [inline client/student/components/preview/common/ContentContainer.vue]: Generalize `ContentContainer` component from `client/student/components/program/Activity` as a common component and reuse on both places.

File: client/student/components/preview/index.vue [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+  },
+  created() {
+    api.getPreview(this.contentPreviewId)
+      .then(activity => {
+        this.activity = activity;
+        this.isLoading = false;
+      });
+  },
+  components: {
+    Activity,
```
Comment (kronicker) [inline client/student/components/preview/index.vue]: Multi-word names ☝️

File: client/student/router.js [hunk 19]
```diff
@@ -67,7 +67,14 @@ const router = new Router({
         props: parseParams
       }]
     }]
-  }, fallbackRoute]
+  },
+  {
+    path: '/preview/:contentPreviewId',
```
Comment (kronicker) [inline client/student/router.js]: `previewId` is verbose enough IMO

File: client/student/router.js [hunk 20]
```diff
@@ -67,7 +67,14 @@ const router = new Router({
...
+    name: 'preview',
+    component: Preview,
+    props: true
```
Comment (kronicker) [inline client/student/router.js]: `parseParams`

File: client/student/components/preview/index.vue [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+
+<script>
+import Activity from './Activity';
+import api from '@/student/api/content';
+import CircularProgress from '@/student/components/common/CircularProgress';
+import Navbar from '@/student/components/common/Navbar';
+
+export default {
+  props: {
+    contentPreviewId: { type: String, required: true }
```
Comment (kronicker) [inline client/student/components/preview/index.vue]: Use `parseParams` in router and use proper Number validation here.

File: server/preview/index.js [hunk 22]
```diff
@@ -0,0 +1,13 @@
+'use strict';
+
+const ctrl = require('./preview.controller');
+const router = require('express').Router();
+
+router
+  .post('/', ctrl.getPreviewUrl)
+  .get('/:id', ctrl.getActivity);
```
Comment (kronicker) [inline server/preview/index.js:10]: Why did you remove cors control for `POST` and `OPTIONS` here?

File: server/preview/preview.model.js [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+      }
+    };
+  }
+
+  static options() {
+    return {
+      tableName: 'preview',
+      modelName: 'preview',
+      timestamps: true,
+      freezesTableName: true
```
Comment (kronicker) [inline server/preview/preview.model.js]: This is redundant if `tableName` is defined

**Review by kronicker — CHANGES_REQUESTED** [review]

File: server/preview/index.js (hunk 22 above)
Comment (zcuric) [reply server/preview/index.js:10]: By mistake.

File: server/preview/preview.model.js (hunk 23 above)
Comment (zcuric) [reply server/preview/preview.model.js]: Need to read `sequelize` docs then.

File: client/student/components/preview/index.vue (hunk 21 above)
Comment (zcuric) [reply client/student/components/preview/index.vue]: OK

File: client/student/components/preview/index.vue (hunk 15 above)
Comment (zcuric) [reply client/student/components/preview/index.vue:49]: More space between content and navbar?

File: client/student/router.js (hunk 19 above)
Comment (zcuric) [reply client/student/router.js]: 👍

File: client/student/router.js (hunk 20 above)
Comment (zcuric) [reply client/student/router.js]: 👍

File: client/student/components/index.vue [hunk 24]
```diff
@@ -17,20 +17,20 @@ import Navbar from '@/student/components/common/Navbar';
...
       .then(programs => {
         setPrograms(programs);
         if (programs.length !== 1 || $route.name === 'activity') return;
         const programId = head(programs).id;
         $router.push({ name: 'courseware', params: { programId } });
       })
-      .finally(() => (this.isLoading = false));
+      .finally(() => {
+        this.isLoading = false;
+      });
```
Comment (kronicker) [inline client/student/components/index.vue]: Inline:
```js
.finally(() => { this.isLoading = false; });
```

**Review by kronicker — CHANGES_REQUESTED** [review]

**Review by kronicker — APPROVED** [review]
